client.go

```diff
@@ -215,7 +240,7 @@ func newBulkPostRequest(urlStr, apiKey string, ips []string) (*http.Request, err
 		return nil, errors.Wrap(err, "failed to encode JSON")
```

Comment:
Now that we have the `ctx`, we should check `if ctx.Err() != nil` before marshaling this JSON. If the context is already canceled no use doing this work.

Are there other places we should check for cancelation now to avoid unnecessary work?

Reply:
Isn't this a bit of premature optimization? JSON _encoding_ isn't quite as bad as _decoding_.

Reply:
I've added an extra check, looks like it's the only place worth checking 🙂

Reply:
@costela _Potentially_. Both encoding and decoding make use of reflection, and it'll have allocated memory for the JSON output that we're discarding. It doesn't seem unreasonable to avoid both cases if we know we're going to throw it all away.